Add a new command to clear harvest_source

kovalch opened the pull request:
To use the cleanup_harvestsources command in the terminal:
```
 docker-compose exec ckan /usr/local/bin/ckan-paster --plugin=ckanext-ogdchcommands ogdch clear_stale_harvestsources --keep_harvestsource_days=30 -c /etc/ckan/production.ini
```

An example output looks like:
```
Cleaning up jobs for all harvest sources
User is authorized to perform this action.
2022-01-14 13:03:40,057 INFO  [ckanext.ogdchcommands.logic] ogdch_cleanup_harvestsource 293 No jobs yet for this harvest source id=da73adbb-ed07-4f6a-bd8a-a1ac96b28e6c
2022-01-14 13:03:40,156 INFO  [ckanext.ogdchcommands.logic] ogdch_cleanup_harvestsource 311 Harvest job id=bf047681-530a-4a99-9f95-b9fa1b370c98 with creation_time=2022-01-14 10:11:35.432290 is not older than 30 days
2022-01-14 13:03:40,230 INFO  [ckanext.ogdchcommands.logic] ogdch_cleanup_harvestsource 293 No jobs yet for this harvest source id=c82a15ce-01e2-4989-9ab7-6a0cf47209ba
2022-01-14 13:03:40,355 INFO  [ckanext.ogdchcommands.logic] ogdch_cleanup_harvestsource 311 Harvest job id=d4961a22-2aca-4bf9-915b-e06b9ee284cd with creation_time=2022-01-14 10:12:07.973740 is not older than 30 days
2022-01-14 13:03:40,430 INFO  [ckanext.ogdchcommands.logic] ogdch_cleanup_harvestsource 293 No jobs yet for this harvest source id=0de7f3f2-4802-4e45-bccc-47172d29565b
```

kovalch committed:
add new command to clear harvest_source

kovalch committed:
add new command to clear harvest_source

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -470,3 +493,79 @@ def _print_shacl_validation_result(self, data_dict):
+        print(data_dict)
+
+        # gets all active harvest sources
+        harvest_sources = model.Session.query(HarvestSource).all()
```
check if `harvest_source_show` can work as well.

kovalch replied:
Use now ```harvest_sources = tk.get_action('harvest_source_list')(context, data_dict)```
instead of ```model.Session```

bellisk committed:
Create LICENSE

kovalch committed:
add working script

kovalch committed:
add working script

kovalch committed:
add cleanup harvestsource func. to the logic

kovalch committed:
add cleanup harvestsource func. to the logic

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -64,6 +64,13 @@ class OgdchCommands(ckan.lib.cli.CkanCommand):
         # that is either today or in the past and sets them to public
         paster ogdch publish_scheduled_datasets [--dryrun]
 
+        # Cleanup harvester sources:
```
I think the purpose of this command is the most important, so maybe you can describe first that this command will check all harvest sources and on basis of the parameter will execute the clearsource command which "deletes all datasets, jobs and objects"

kovalch replied:
Changed to:
``` 
# Cleanup harvester sources:
# - check all harvest sources and the presents of harvest jobs,
#   for harvesters, which last jobs are finished for more than n-days,
#   the clearsource command will be executed
#   that deletes all datasets, jobs and objects, but keeps the source itself
# - the default timeframe to keep harvested datasets is 30 days
```

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -64,6 +64,13 @@ class OgdchCommands(ckan.lib.cli.CkanCommand):
+        # - deletes all datasets, jobs and objects related to a harvest source,
+        #     but keeps the source itself
+        # - the default timeframe to keep harvested datasets of a harvest-source: 30 days
+        paster ogdch cleanup_harvestsources
```
Consider naming the command `clear_stale_harvestsources`, maybe that way it is easier to understand immediately, what the purpose of this command is and also shows immediately that it is based on `clearsource`

kovalch replied:
use a recommended name for the command `clear_stale_harvestsources `

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -83,6 +90,10 @@ def __init__(self, name):
+        self.parser.add_option(
+            '--keep_harvestsource', action="store", type="int", dest='tf_to_keep_harvested_dsets',
+            default=30,
+            help='Initial Timeframe to keep harvested datasets')
```
Consider adding `.. to keep harvested datasets, jobs and objects.`.

kovalch replied:
done

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -83,6 +90,10 @@ def __init__(self, name):
             default='ech-0200.shacl.ttl',
             help='shape file name for shacl shape validation')
+        self.parser.add_option(
+            '--keep_harvestsource', action="store", type="int", dest='tf_to_keep_harvested_dsets',
```
This is probably more a taste or my personal preference but I would rather not use abbreviations unless we have to. As there is a line-break here, anyways, the words can also be spelled out, IMHO
But if you think it makes it easier to read this way, that is okay, too :)

kovalch replied:
Remove abbreviations: `tf_to_keep_harvested_dsets` to `timeframe_to_keep_harvested_datasets`

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -470,3 +482,44 @@ def _print_shacl_validation_result(self, data_dict):
+
+    def cleanup_harvestsources(self, source=None):
+        """
+        command that clear all datasets, jobs and objects related to a harvest source
```
clears

kovalch replied:
done

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -470,3 +482,44 @@ def _print_shacl_validation_result(self, data_dict):
+    def cleanup_harvestsources(self, source=None):
+        """
+        command that clear all datasets, jobs and objects related to a harvest source
+        that was not active for the last 30 days. 30 days is the default value,
```
Consider: ... that was not active for a given amount of days (default 30 days).

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -470,3 +482,44 @@ def _print_shacl_validation_result(self, data_dict):
+        """
+        command that clear all datasets, jobs and objects related to a harvest source
+        that was not active for the last 30 days. 30 days is the default value,
+        use --keep_harvestsource_timeframe=n to change timeframe of keeping harvester resources.
```
now that I read timeframe, I am wondering if naming it `keep_harvestsource_days` might be more straight forward. What do you think?

harvester resources -> harvester objects (either name all three: jobs, datasets, objects or just objects to stay generic enough)

kovalch replied:
change `keep_harvestsource_timeframe` to `keep_harvestsource_days`
and use  `harvester objects ` for  generic description

stefina commented on ckanext/ogdchcommands/commands.py:
```diff
@@ -470,3 +482,44 @@ def _print_shacl_validation_result(self, data_dict):
+        # get source from arguments
+        source_id = None
+        data_dict = {}
+        if len(self.args) >= 2:
```
It is nice to create a nice output but I wouldn't make a request just for a nice output :) (If you get the info for free somewhere else, that would be different, IMHO)
To me it would be okay if it just says plural and would only check one source if there were only one.

kovalch replied:
left only `Cleaning up jobs for all harvest sources` output

stefina commented on ckanext/ogdchcommands/logic.py:
```diff
@@ -269,3 +270,47 @@ def ogdch_cleanup_resources(context, data_dict):
+            last_job_creation_time_obj = datetime.datetime.strptime(last_job_creation_time,
+                                                                    "%Y-%m-%d %H:%M:%S.%f")
+
+            if (last_job_creation_time_obj < last_day_to_keep_harvested_ds):
```
I am wondering if it makes sense to also check if the latest job is finished (`status`) not sure if we want to delete jobs that are probably stuck (even if it is for those many days) with force. WDYT?

kovalch replied:
You are right, especially if someone wants to run current command for harvesters with last job age of, e.g., 5 days.
Added this check

stefina commented on ckanext/ogdchcommands/logic.py:
```diff
@@ -269,3 +270,47 @@ def ogdch_cleanup_resources(context, data_dict):
+        if not source_dict['status']['last_job']:
+            log.info('No jobs yet for this harvest source id={}'.format(source['id']))
+        else:
+            jast_job_id = source_dict['status']['last_job']['id']
```
Typo? `last_job_id`

kovalch replied:
yes, corrected it

stefina commented on ckanext/ogdchcommands/logic.py:
```diff
@@ -269,3 +270,47 @@ def ogdch_cleanup_resources(context, data_dict):
+
+            if (last_job_creation_time_obj < last_day_to_keep_harvested_ds):
+                log.info('Harvest latest job id={} with creation_time={} is older than {} days'
+                      .format(jast_job_id, last_job_creation_time,
+                              tf_to_keep_harvested_dsets))
+                log.info('Clears all datasets, jobs and objects related to a harvest source id={}'
+                      .format(jast_job_id))
+                count_cleared_harvestsource += count_cleared_harvestsource
+                tk.get_action("harvest_source_clear")(context, {"id": source['id']})
```
IMHO you can reduce the output a little. The command itself and the documentation states what this command does, so I see no need to describe it again for every cleared source.
I might be more useful to give an output of how old the latest job was.
So consider if all of this as output is really useful.

kovalch replied:
This is how the new output will look like:
```
User is authorized to perform this action
Cleaning up harvester objects for all harvest sources
2022-01-17 09:25:03,388 INFO  [ckanext.ogdchcommands.logic] ogdch_cleanup_harvestsource 303 The latest job of the harvester id=da73adbb-ed07-4f6a-bd8a-a1ac96b28e6c is 0 days old
2022-01-17 09:25:03,545 DEBUG [ckanext.harvest.logic.action.update] harvest_source_reindex 936 Updating search index for harvest source: zh-kt-gemeindeamt-dcat-harvester
2022-01-17 09:25:03,694 INFO  [ckanext.ogdchcommands.logic] ogdch_cleanup_harvestsource 303 The latest job of the harvester id=113ecdc2-134c-4068-b4af-4cd1e9a44920 is 0 days old
2022-01-17 09:25:03,794 DEBUG [ckanext.harvest.logic.action.update] harvest_source_reindex 936 Updating search index for harvest source: zh-kt-volksschulamt-dcat-harvester
2022-01-17 09:25:03,944 INFO  [ckanext.ogdchcommands.logic] ogdch_cleanup_harvestsource 295 No jobs yet for this harvest source id=c82a15ce-01e2-4989-9ab7-6a0cf47209ba
2 harvest sources were cleared
```

Commit:
remove the command 'shacl_validate'

this command is not used any more. Shacl validation is instead performed
via the shacl checker that resides in the repo https://github.com/opendata-swiss/ogdch_checker

Commit:
reformat some exception and print statements

Commit:
remove unnessessary dependencies

Commit:
remove unneccessary parts from the build

Commit:
add command to clean up a key in package_extra

the command has a dryrun option: --dryrun
it deletes all entries for a key in the tables package_extra and
the dependent table package_extra_revision

Commit:
remove documentation of the shacl validate command in the README

Commit:
add documentation for the command cleanup_extra in the README

kovalch committed:
implementing dev review comments from Steffi

kovalch committed:
implementing dev review comments from Steffi

sabinem committed:
Merge pull request #8 from opendata-swiss/remove-shacl-validate-command

remove the command 'shacl_validate'

sabinem committed:
Update README.md

Co-authored-by: Rae Knowler <bellisk@users.noreply.github.com>

sabinem committed:
Update ckanext/ogdchcommands/commands.py

Co-authored-by: Rae Knowler <bellisk@users.noreply.github.com>

Commit:
correct the description of the cleanup_extras command in the code

Commit:
add cleaup_extras command to the helptext for dryrun option

sabinem committed:
Merge pull request #9 from opendata-swiss/add-package-extra-cleanup-command

Add package extra cleanup command

kovalch committed:
updated command file after using rebase to the latest master

kovalch committed:
update files after using rebase to the latest master

kovalch committed:
fixed the code style

kovalch committed:
fixed the code style#2

kovalch committed:
remove blank lines at end of files
